rust/rebase/src/signer/signer.rs

```diff
@@ -56,11 +56,25 @@ pub struct EIP155 {
     pub chain_id: String,
 }
 
+#[derive(Clone, Deserialize, JsonSchema, Serialize)]
+#[serde(rename = "solana")]
+pub struct Solana {
+    pub address: String,
+}
+
+// TODO: Understand where this came from.
+// Noted in the did:pkh w3 docs as the network for did:pkh:solana
+// but should note where / how that's discoverable to know if it's going to change
+// or if it's not going to change, then note it as a magic string.
+pub const SOLANA_NETWORK: &str = "4sGjMW1sUnHzSxGspuhpqLDx6wiyjNtZ";
+
 #[derive(Clone, Deserialize, JsonSchema, Serialize)]
 #[serde(rename = "pkh")]
 pub enum PKH {
     #[serde(rename = "eip155")]
     EIP155(Option<EIP155>),
+    #[serde(rename = "solana")]
+    Solana(Option<Solana>),
```

Comment:
why the optional?

Reply:
So you can supply a public key at runtime -- so you can state what signature type _will_ be there in the future, then later supply the actual public key material.

Reply:
I find this confusing, this is the signer for the VCs right? I would personally prefer two different config variables: one for the JWK and the other for the DID

Reply:
I don't know what you mean. You feed a JWK and a did identifier in when you create a new Signer (specifically, the Ed25519DidWebJwk variant). If the JWK doesn't line up to the did identifier (meaning, if the URL failed to resolve to the public key), it will fail on issuance.  What should happen differently?

Reply:
Apologies for the brain dump:
1. If issuers are always DIDs, I don't understand the need for this `SignerType` trait (which should probably be named something like `VerifyingIssuer`) -- you can resolve the DID with ssi (and it will give you the pk JWK if it can) and also the signature verification (and again, it might be able to do it without the need of a JWK).
2. The `Signer` feels like it's already in ssi or it should be.
3. What's a `did_id`?
4. I don't think embedding traits in traits (`SignerType` in `Signer`) is usually a good idea. `ssi` does that but I think it's rubbish.

Reply:
Summary of the brain dump: I think we should be able to completely rely on `ssi` for the crypto side of things. It will probably require work on `ssi` though, as its interface is probably incomplete but also scattered.

Reply:
1) SignerTypes aren't only used for VCs. They are primarily used with Plain Text. Do you remember how TZProfiles worked? That Tezos wallet would sign in plain text and witness would verify it? Rebase is that flow abstracted, so we need an abstraction that covers verification of a Plain Text signature and Public Key. I don't think DIDKit or SSI provide that.
2) That's interesting, what do you mean?
3) `did()` returns a `DID` struct. `did_id()` returns a string that is the DID (meaning `did:pkh:...`/`did:web:...`). Do you think there's a better way to name these?
4) Why is it rubbish?
Summary: If SSI worked, I would use it, but I need meet deadlines.

Reply:
Another consideration for Rebase's Signer / SignerType is that the traits are used to fill out the credential at runtime. Does SSI do that?

Reply:
I took a long look at SSI to try to understand what you're getting at, and once SSI is actually working and published to crates.io, one could open an issue on Rebase attempting to accomplish the following:

1. Change `Signer<T: SignerType>` to `Signer where Self: SignerType`. 
2. Add the following methods to SignerType:
`fn jwk(&self) -> JWK`
`fn algo(&self) -> Algorithm`.
3. Rework the functions to create a DID (the string identifier (did:web/pkh:...), not the document) and remove the DID struct.
3. Change signing / verification to use:
`&self.jwk().<sign/sign_bytes/verify/verify_bytes>()`

There are some serious trade-offs to this. Since `SignerType` only represents a Public Key and `Signer` represents a Public/Private key pair, it is possible to construct an invalid `Signer` by not including a private key in the JWK. It becomes possible with both structs to improperly assign the DID (Ed25519 needs to be distinguished between `did:web` and `did:pkh:solana`, etc).

Additionally, because Ethereum's cryptography isn't particularly strong, we did not want to support VCs issued by Ethereum accounts, but we did want the ability to witness claims signed by Ethereum accounts. Additional handling would be required to make sure only a subset of Algorithms were supported for `Signer`s versus `SignerType`s. At the present, you can make this distinction by only implementing a `SignerType` for a given key pair type.

These mistakes are currently compile-time errors, but would become runtime errors, and would likely surface very far from where the actual mistake occurred. If you see advantages I'm missing an believe that this would be a better solution, I encourage you to open this issue once SSI's new structure is published to crates.io and we have a firm foundation to work off of. It would help if you concretely explained your perspective to highlight advantages this approach would enable.

Either way, I am very curious why the pattern `OuterTrait<T: InnerTrait>` is rubbish. An explanation would add a constructive element to that criticism.